[add-friends-likes] add likes and friends, add storage and service layers

PascalABCD opened the pull request.

pascalplata committed:
[add-friends-likes] add likes and friends, add storage and service layers

pascalplata committed:
[add-friends-likes] fix style

pascalplata committed:
[add-friends-likes] fix style

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,91 @@
+    public Film update(Film film) {
+        return filmStorage.update(film);
+    }
+
```
напишу одним комментом для всех случаев проверки существования фильма и пользователя. Заготовь в интерфейсах  FilmStorage и UserStorage методы getById(). А в текущем классе добавь два внутренних метода, которые вызовет этот getById(), проверит на null (если да, бросит исключение с комментом) и вернет соответствующую сущность. Эти методы позволят проверить существование фильма/пользователя не повторяя один и тот же код

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,91 @@
+                .findFirst()
+                .orElseThrow(() -> new NotFoundException("Фильм не найден"));
+
+        if (film.getLikes().contains(userId)) {
```
в фильме в хранилище likes хранятся пользователи объектом, но лучше переделать на id

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public List<Film> getPopularFilms(int count) {
+        return filmStorage.getAll().stream()
```
это не обязательный комментарий, но я бы сортировку и limit перенес в InMemoryFilmStorage. Поясню причину. Пока хранилище мапа, разницы нет где проводятся эти процедуры. Но после смены мапы на БД, придется точно переписывать, так как тащить все и только потом отсекать не нужное - ужасная практика. Все БД умеют сортировать и возвращать только нужное количество элементов. А это значит потом нужно будет править и сервисный слой. А разработку нужно вести так, чтобы замена одной реализации DAO-прослойки никак не влияла на сервисный слой

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    public User create(User user) {
+        if (user.getName() == null || user.getName().isBlank()) {
```
вынеси логику переноса логина в имя во внутренний метод класса и переиспользуй

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,135 @@
+        log.info("Пользователь {} обновлен", user);
+        return storage.update(user);
+    }
+
```
в этом классе добавь аналогичный метод получения пользователя или выброс исключения

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,135 @@
+                    return new NotFoundException("Пользователь не найден");
+                });
+        log.info("Пользователь {} найден", userId);
+        return storage.getAll().stream()
```
здесь нужно у user взять список друзей и только их собрать в список, а не перебирать все хранилище

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,135 @@
+                });
+
+        log.info("Выводим список общих друзей пользователей {} и {}", userId, otherId);
+        return storage.getAll().stream()
```
здесь нужно сначала сопоставить списки друзей у пользователей и только потом по этому результирующему списку собрать список пользователей

pascalplata committed:
[add-friends-likes] review fixes:

- add getById() for films and users
- move sorting and limit to storage
- add checkName
- other fixes

pascalplata committed:
[add-friends-likes] fix

pascalplata committed:
[add-friends-likes] fix
